accelerator/common_core.c

```diff
@@ -284,68 +284,88 @@ status_t ta_find_transactions_by_tag(const iota_client_service_t* const service,
-    ret = SC_CCLIENT_OOM;
-    goto done;
+  flex_trit_t tx_trits[NUM_TRITS_SERIALIZED_TRANSACTION];
+  iota_transaction_t* temp = NULL;
+  get_trytes_req_t* req_get_trytes = get_trytes_req_new();
+  transaction_array_t* uncached_txn_array = transaction_array_new();
+  if (req == NULL || res == NULL || req_get_trytes == NULL || uncached_txn_array == NULL) {
+    return SC_TA_NULL;
   }
+  char txn_hash[NUM_TRYTES_HASH + 1] = {0};
+  char cache_value[NUM_TRYTES_SERIALIZED_TRANSACTION + 1] = {0};
+  txn_hash[NUM_TRYTES_HASH] = '\0';
+  cache_value[NUM_TRYTES_SERIALIZED_TRANSACTION] = '\0';
 
-  // get in cache first, then get from full node if no result in cache
-  ret = cache_get(req, cache_value);
-  if (ret == SC_OK) {
-    flex_trits_from_trytes(tx_trits, NUM_TRITS_SERIALIZED_TRANSACTION, (const tryte_t*)cache_value,
-                           NUM_TRYTES_SERIALIZED_TRANSACTION, NUM_TRYTES_SERIALIZED_TRANSACTION);
-  } else {
-    ret = hash243_queue_push(&get_trytes_req->hashes, hash_trits);
-    if (ret) {
-      ret = SC_CCLIENT_HASH;
-      goto done;
+  // append transaction object which is already cached to transaction_array_t
+  // if not, append uncached to request object of `iota_client_find_transaction_objectss`
+  hash243_queue_entry_t* q_iter = NULL;
+  CDL_FOREACH(req->hashes, q_iter) {
+    flex_trits_to_trytes((tryte_t*)txn_hash, NUM_TRYTES_HASH, q_iter->hash, NUM_TRITS_HASH, NUM_TRITS_HASH);
+
+    ret = cache_get(txn_hash, cache_value);
+    if (ret == SC_OK) {
+      flex_trits_from_trytes(tx_trits, NUM_TRITS_SERIALIZED_TRANSACTION, (const tryte_t*)cache_value,
+                             NUM_TRYTES_SERIALIZED_TRANSACTION, NUM_TRYTES_SERIALIZED_TRANSACTION);
+
+      // deserialize raw data to transaction object
+      temp = transaction_deserialize(tx_trits, true);
+      if (temp == NULL) {
+        ret = SC_CCLIENT_OOM;
+        goto done;
+      }
+
+      transaction_array_push_back(res, temp);
+      transaction_free(temp);
+    } else {
+      if (hash243_queue_push(&req_get_trytes->hashes, q_iter->hash) != RC_OK) {
+        ret = SC_CCLIENT_OOM;
+        goto done;
+      }
+      ret = SC_OK;
     }
 
-    ret = iota_client_get_trytes(service, get_trytes_req, get_trytes_res);
-    if (ret) {
+    // reset the string `cache_value`
+    cache_value[0] = '\0';
+  }
+
+  if (req_get_trytes->hashes != NULL) {
+    if (iota_client_get_transaction_objects(service, req_get_trytes, uncached_txn_array) != RC_OK) {
       ret = SC_CCLIENT_FAILED_RESPONSE;
       goto done;
     }
+  }
 
-    memcpy(tx_trits, hash8019_queue_peek(get_trytes_res->trytes), FLEX_TRIT_SIZE_8019);
-
-    // set into cache if get_trytes response is not null trytes
-    if (!flex_trits_are_null(tx_trits, FLEX_TRIT_SIZE_8019)) {
-      flex_trits_to_trytes((tryte_t*)cache_value, NUM_TRYTES_SERIALIZED_TRANSACTION, tx_trits,
+  // append response of `iota_client_find_transaction_objectss` into cache
+  flex_trit_t* temp_txn_trits = NULL;
```

Comment:
`temp_txn_trits` could be moved into `TX_OBJS_FOREACH` loop, since it only used here.
And the freed opearation could be removed in `done`

Reply:
If I move the declaration of `temp_txn_trits`, it will take time to allocate space for this variable each time in this loop. I think it may be not effective.
The other thing is the function `transaction_serialize()` would create a new `flex_trit_t*` object with `malloc`, so we need to free the `flex_trit_t*` object each time in loop.
Therefore I prefer the current way.